compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -427,4 +476,386 @@ LogicalResult splitLogicalObjectFifos(
+    return newL2ToL1DmaOp;
+  };
+  for (unsigned i = 0, n = l3ToL2DmaOps.size(); i < n; i += 2) {
+    // Step 1. Combine the picked L3->L2 DmaCpyNd pair.
+    FailureOr<LogicalObjectFifoFromMemrefOp> maybeNewL2ObjectFifo =
+        combineL3ToL2Pair(rewriter, l3ToL2DmaOps[i], l3ToL2DmaOps[i + 1],
+                          splitDims, nonSplitDims);
+    if (failed(maybeNewL2ObjectFifo)) return failure();
+    LogicalObjectFifoFromMemrefOp newL2ObjectFifo =
+        maybeNewL2ObjectFifo.value();
+
+    // Step 2. We now have need to create two L2->L1 ops since the size has
+    // changed. But for this we first need to find the new offset for L2 as
+    // source.
+    // TODO: For now I'm hardcoding the offsets but later it'd just depend
+    // on split/non-split dimensions.
+    // Offset = 0,0
+    LogicalObjectFifoFromMemrefOp reuseL1LogicalObjectFifoOp =
+        l2ToL1DmaOps[i].getTargetObjectFifo();
+    SmallVector<OpFoldResult> newL2AsSourceOffsets =
+        l2ToL1DmaOps[i].getSourceMixedOffsets();
+    DmaCpyNdOp newFirstL2ToL1DmaOp = createL2ToL1ForReuse(
+        rewriter, l2ToL1DmaOps[i], reuseL1LogicalObjectFifoOp, newL2ObjectFifo,
+        newL2AsSourceOffsets);
+    // Offset = 0, 1. NOTE here we'd use the same L1 logical objectFifo as
+    // the first L2->L1 Dma.
+    newL2AsSourceOffsets = l2ToL1DmaOps[i + 1].getSourceMixedOffsets();
+    newL2AsSourceOffsets[1] = rewriter.getIndexAttr(1);
+    DmaCpyNdOp newSecondL2ToL1DmaOp = createL2ToL1ForReuse(
+        rewriter, l2ToL1DmaOps[i + 1], reuseL1LogicalObjectFifoOp,
+        newL2ObjectFifo, newL2AsSourceOffsets);
+
+    // Step 3. PICK the CoreOps associated with the 1:1 L2->L1.
+    // For the first Core op we'll insert Read at the end. It doesn't matter
+    // for now so we're gonna insert it right before amdaie.end op.
+    CoreOp firstCoreOp = fetchUniqueCoreOp(newFirstL2ToL1DmaOp);
+    firstCoreOp.walk([&](AMDAIE::EndOp endOp) {
+      OpBuilder::InsertionGuard guard(rewriter);
+      // Hardcoding to `AMDAIE::MemoryAccess::Read`.
+      rewriter.setInsertionPoint(endOp);
+      rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+          rewriter.getUnknownLoc(), reuseL1LogicalObjectFifoOp.getOutput(),
+          AMDAIE::MemoryAccess::Read);
+    });
+    // For the second Core op we'll insert `Read` right before the first read
+    // from the corresponding L1 logicalobjectFifo.
+    CoreOp secondCoreOp = fetchUniqueCoreOp(newSecondL2ToL1DmaOp);
+    secondCoreOp.walk([&](AMDAIE::LogicalObjectFifoAccessOp accessOp) {
+      if (accessOp.getInput() == l2ToL1DmaOps[i + 1].getTargetObjectFifo()) {
+        OpBuilder::InsertionGuard guard(rewriter);
+        // Hardcoding to `AMDAIE::MemoryAccess::Read`.
+        rewriter.setInsertionPoint(accessOp);
+        rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+            rewriter.getUnknownLoc(), reuseL1LogicalObjectFifoOp.getOutput(),
+            AMDAIE::MemoryAccess::Read);
```

Comment:
You can derive whether it's a read or write from whether it's an input/output DMA to the core.